[java] NPE in type inference

**Affects PMD Version:** 
7.5.0 (this didn't happen with version 7.4.0 and under)

**Rule:** 
No rules as its failing to parse the code

**Description:**
```
 [INFO] --- pmd:3.25.0:pmd (pmd) @ uk-multi-policy-group-service ---
17:53:18   [INFO] PMD version: 7.5.0
17:53:27   [ERROR] PMD processing errors:
17:53:27   [ERROR] /home/teamcity/work/402506fd6a30bd90/uk-multi-policy-group-service/src/test/java/com/marshmallow/uk_multi_policy_group/persistence/policy_group_numbers/PolicyGroupNumberGeneratorStoreTest.java: ContextedRuntimeException: java.lang.NullPointerException: Cannot invoke "net.sourceforge.pmd.lang.java.types.JTypeMirror.equals(Object)" because "asSuperOfS" is null
17:53:27   Exception Context:
17:53:27     [1:Resolving type of=!debug only! [MethodCall:174:16]Optional.ofNullable(
17:53:27               Iterables.getOnlyElement(queryForKey(key), null)
17:53:27           )]
17:53:27     [2:Location=!debug only! /home/teamcity/work/402506fd6a30bd90/uk-multi-policy-group-service/src/test/java/com/marshmallow/uk_multi_policy_group/persistence/policy_group_numbers/PolicyGroupNumberGeneratorStoreTest.java:174:16]
17:53:27   ---------------------------------
```

**Code Sample demonstrating the issue:**

```java
    private Optional<Map<String, AttributeValue>> loadForKey(final String key) {
        return Optional.ofNullable(
            Iterables.getOnlyElement(queryForKey(key), null)
        );
    }
```

**Expected outcome:**
PMD reports a violation at line ..., but that's wrong. That's a false positive.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*
Maven (v3.5.0)

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/TypeSystem.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/TypeSystem.java
@@ -799,6 +799,11 @@ public final class TypeSystem {
         }
 
         @Override
+        public @Nullable JTypeMirror getAsSuper(@NonNull JClassSymbol symbol) {
+            throw new UnsupportedOperationException("Null type cannot call asSuper, will return null always");
+        }
+
+        @Override
         public <T, P> T acceptVisitor(JTypeVisitor<T, P> visitor, P p) {
             return visitor.visitNullType(this, p);
         }

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/internal/infer/Infer.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/internal/infer/Infer.java
@@ -816,7 +816,7 @@ public final class Infer {
 
     private boolean commonSuperWithDiffParameterization(JTypeMirror t, JTypeMirror s) {
         JTypeMirror lubResult = ts.lub(listOf(t, s));
-        if (lubResult.isBottom() || lubResult.isTop()) {
+        if (lubResult.isBottom() || lubResult.isTop() || t.isBottom() || s.isBottom()) {
             return false;
         }
         for (JTypeMirror sup : asList(lubResult)) {
@@ -824,6 +824,8 @@ public final class Infer {
                 JClassSymbol sym = ((JClassType) sup).getSymbol();
                 JTypeMirror asSuperOfT = t.getAsSuper(sym);
                 JTypeMirror asSuperOfS = s.getAsSuper(sym);
+                assert asSuperOfS != null : "s <: sup, because sup is part of the LUB of s";
+                assert asSuperOfT != null : "t <: sup, because sup is part of the LUB of t";
                 if (!asSuperOfS.equals(asSuperOfT)) {
                     return true;
                 }

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/internal/infer/InferenceVar.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/internal/infer/InferenceVar.java
@@ -97,7 +97,6 @@ public final class InferenceVar implements SubstVar {
      * Adds a new bound on this variable.
      */
     public void addBound(BoundKind kind, JTypeMirror type) {
-        this.hasNonTrivialBound = true;
         addBound(kind, type, false);
     }
 
@@ -115,9 +114,15 @@ public final class InferenceVar implements SubstVar {
             // may occur because of transitive propagation
             // alpha <: alpha is always true and not interesting
             return;
+        } else if (kind == BoundKind.LOWER && type.isBottom()) {
+            // null <: alpha is not interesting and may cause errors because of lub.
+            return;
         }
 
         if (boundSet.bounds.computeIfAbsent(kind, k -> new LinkedHashSet<>()).add(type)) {
+            if (!isPrimaryBound) {
+                this.hasNonTrivialBound = true;
+            }
             ctx.onBoundAdded(this, kind, type, isPrimaryBound);
         }
     }

--- a/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/types/internal/infer/InferenceCtxUnitTests.java
+++ b/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/types/internal/infer/InferenceCtxUnitTests.java
@@ -74,6 +74,22 @@ class InferenceCtxUnitTests extends BaseTypeInferenceUnitTest {
         verify(log).boundAdded(ctx, v2, BoundKind.UPPER, listOfV1, false);
     }
 
+
+    @Test
+    void testNullTypeCannotBeLowerBound() {
+        TypeInferenceLogger log = spy(TypeInferenceLogger.noop());
+        InferenceContext ctx = emptyCtx(log);
+
+        InferenceVar v1 = newIvar(ctx);
+
+        addSubtypeConstraint(ctx, ts.NULL_TYPE, v1);
+
+
+        assertThat(v1, hasBoundsExactly(upper(ts.OBJECT)));
+
+        verify(log, never()).boundAdded(ctx, v1, BoundKind.LOWER, ts.NULL_TYPE, false);
+    }
+
     @Test
     void testEqBoundWithGenerics() {
         TypeInferenceLogger log = spy(TypeInferenceLogger.noop());

--- a/pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/types/internal/infer/TypeInferenceTest.kt
+++ b/pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/types/internal/infer/TypeInferenceTest.kt
@@ -454,4 +454,32 @@ public class BadIntersection {
             info.methodType.formalParameters[0] shouldBe optOutEnum
         }
     }
+    parserTest("#5190 NPE in type inf") {
+        val (acu, spy) = parser.parseWithTypeInferenceSpy(
+            """
+    interface Optional<T> {
+        static Optional<T> ofNullable(T t) {}
+    }
+    interface Map<K,V> {}
+    interface AttributeValue{}
+    public class Main {
+        private Optional<Map<String, AttributeValue>> loadForKey(final String key) {
+            return Optional.ofNullable(
+                Iterables.getOnlyElement(queryForKey(key), null)
+            );
+        }
+    }
+
+            """.trimIndent()
+        )
+
+        val (_, _, _) = acu.declaredTypeSignatures()
+        val (ofNullable) = acu.methodDeclarations().toList { it.genericSignature }
+
+        spy.shouldBeOk {
+            val info = acu.firstMethodCall().overloadSelectionInfo
+            info::isFailed shouldBe false
+            info.methodType shouldBeSomeInstantiationOf ofNullable
+        }
+    }
 })

--- a/pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/types/internal/infer/TypeInferenceTest.kt
+++ b/pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/types/internal/infer/TypeInferenceTest.kt
@@ -457,19 +457,35 @@ public class BadIntersection {
     parserTest("#5190 NPE in type inf") {
         val (acu, spy) = parser.parseWithTypeInferenceSpy(
             """
-    interface Optional<T> {
-        static Optional<T> ofNullable(T t) {}
+    import java.util.Iterator;
+    interface Optional<V> {
+        static <T> Optional<T> ofNullable(T t) {}
     }
     interface Map<K,V> {}
+    interface List<V> extends Iterable<V> {}
     interface AttributeValue{}
-    public class Main {
+    
+class Main {
+
         private Optional<Map<String, AttributeValue>> loadForKey(final String key) {
             return Optional.ofNullable(
-                Iterables.getOnlyElement(queryForKey(key), null)
+                getOnlyElement(queryForKey(key), null)
             );
         }
-    }
 
+        private List<Map<String, AttributeValue>> queryForKey(final String key) {
+            return null;
+        }
+
+        public static <T> T getOnlyElement(final Iterable<? extends T> iterable, final T defaultValue) {
+            return getOnlyElement(iterable.iterator(), defaultValue);
+        }
+
+        public static <T> T getOnlyElement(final Iterator<? extends T> iterator, final T defaultValue) {
+            return null;
+        }
+
+    }
             """.trimIndent()
         )
 

